fusequery/query/src/pipelines/transforms/transform_group_by_final.rs

```diff
@@ -86,112 +82,156 @@ impl Processor for GroupByFinalTransform {
+                        values.push(vec![])
+                    }
+                    values
+                };
+
+                let mut aggr_values: Vec<Vec<DataValue>> = {
+                    let mut values = vec![];
+                    for _i in 0..aggr_funcs_len {
+                        values.push(vec![])
+                    }
+                    values
+                };
+                for (_key, (places, values)) in groups.iter() {
+                    for (i, value) in values.iter().enumerate() {
+                        group_values[i].push(value.clone());
+                    }
+
+                    for (i, func) in aggr_funcs.iter().enumerate() {
+                        let merge = func.merge_result(places[i])?;
+                        aggr_values[i].push(merge);
+                    }
                 }
-            }
 
-            for (i, func) in aggr_funcs.iter().enumerate() {
-                let merge = func.merge_result()?;
-                aggr_values[i].push(merge);
-            }
-        }
+                // Build final state block.
+                let mut columns: Vec<Series> = Vec::with_capacity(aggr_funcs_len + group_expr_len);
 
-        // Build final state block.
-        let mut columns: Vec<Series> = Vec::with_capacity(aggr_funcs_len + group_expr_len);
+                for (i, value) in aggr_values.iter().enumerate() {
+                    columns.push(DataValue::try_into_data_array(
+                        value.as_slice(),
+                        &self.aggr_exprs[i].to_data_type(&self.schema_before_group_by)?,
+                    )?);
+                }
 
-        for value in &aggr_values {
-            if !value.is_empty() {
-                columns.push(DataValue::try_into_data_array(value.as_slice())?);
-            }
-        }
+                for (i, value) in group_values.iter().enumerate() {
+                    columns.push(DataValue::try_into_data_array(
+                        value.as_slice(),
+                        &self.group_exprs[i].to_data_type(&self.schema_before_group_by)?,
+                    )?);
+                }
 
-        for value in &group_values {
-            if !value.is_empty() {
-                columns.push(DataValue::try_into_data_array(value.as_slice())?);
-            }
+                let mut blocks = vec![];
+                if !columns.is_empty() {
+                    let block = DataBlock::create_by_array(self.schema.clone(), columns);
+                    blocks.push(block);
+                }
+
+                Ok(Box::pin(DataBlockStream::create(
+                    self.schema.clone(),
+                    None,
+                    blocks,
+                )))
+            }};
         }
 
-        let mut blocks = vec![];
-        if !columns.is_empty() {
-            let block = DataBlock::create_by_array(self.schema.clone(), columns);
-            blocks.push(block);
+        macro_rules! match_hash_method_and_apply {
```

Comment:
neat

Reply:
Yes, but It's a little hard to make it more generic rather than macros.